scinet/scinet.py

```diff
@@ -47,73 +50,118 @@ def __init__(self, hyp):
+        # Appy formula for Z
+        Z = mu + gauss * std
+        # Return them
+        return Z
+
+    def decode(self, Z, question):
         # Concatenate output of encoder with question
-        x = torch.cat((question, x), dim=-1)
+        Z = torch.cat((question, Z), dim=-1)
 
         # Pass through decoder layers (without applying relu on answer neuron)
         lastDecoderLayer = len(self.decoder) - 1
 
         for i, layer in enumerate(self.decoder):
-            x = layer(x)
+            Z = layer(Z)
 
             if i != lastDecoderLayer:
-                x = funct.relu(x)
+                Z = funct.relu(Z)
+        return Z
+
+    def forward(self, x, question, verbose=False):
 
-        return x
+        # pass through encoder
+        mu, sig = self.encode(x)
 
-    def train(self, observations, questions, answers, batchSize):
+        # reparamaterize
+        Z = self.reparameterize(mu, sig)
+
+        # decode
+        Y = self.decode(Z, question)
+
+        return mu, sig, Z, Y
+
+    def train(self, observations, questions, answers, batchSize, verbose=False):
 
         avgLoss = 0
         trainSize = len(observations)
 
-        for i in tqdm(range(0, trainSize, batchSize)):
+        for i in range(0, trainSize, batchSize):
 
             observationBatch = observations[i:i + batchSize].to(self.device)
             answersBatch = answers[i:i + batchSize].to(self.device)
             questionBatch = questions[i:i + batchSize].to(self.device)
 
             self.zero_grad()
-            outputs = self(observationBatch, questionBatch)
-            loss = self.lossFunct(outputs, answersBatch)
+            mu, sig, Z, outputs = self(observationBatch, questionBatch, verbose=verbose)
+            loss = self.lossFunct(mu, sig, outputs, answersBatch)
             loss.backward()
             self.optimizer.step()
 
             avgLoss += loss.item() * len(observationBatch)
 
         avgLoss /= trainSize
-        print("Training loss:", avgLoss)
+
+        if verbose:
+            print("Training loss:", avgLoss)
+
         return (avgLoss)
 
-    def test(self, observations, questions, answers, batchSize):
+    def test(self, observations, questions, answers, batchSize, verbose=False):
 
         avgLoss = 0
         testSize = len(observations)
 
         with torch.no_grad():
-            for i in tqdm(range(0, testSize, batchSize)):
+            # for i in tqdm(range(0, testSize, batchSize)):
+            for i in range(0, testSize, batchSize):
 
                 observationBatch = observations[i:i + batchSize].to(self.device)
                 answersBatch = answers[i:i + batchSize].to(self.device)
                 questionBatch = questions[i:i + batchSize].to(self.device)
 
-                outputs = self(observationBatch, questionBatch)
-                loss = self.lossFunct(outputs, answersBatch)
+                mu, sig, Z, outputs = self(observationBatch, questionBatch, verbose=verbose)
```

Comment:
How are you guys normally inspecting the latent neuron activations? I'm realizing now that in neither of these versions is the activation (Z) actually returned, in any way.

In `copernicus` (in the test method) I'm saving the latent activations as we train, then returning them alongside the loss, to analyze. Should we be doing something like that here as well? I'm just curious what you all are doing right now, without that.

Reply:
@keyblaede

Reply:
I have not been examining activation as my results haven't been strong enough to bother yet. I am not sure how to go about doing this, but I an down to do it any way you would like and I can implement it if you would like.

Reply:
@nmdickso

Reply:
Okay, don't worry about it, I'll work with Veronica to implement something, probably in a similar fashion to how I've done it for `copernicus`, it's mostly just for visualizations anyways.
Thanks